### ISSUE TITLE
zlint 3.2.0 can't compile as a library

### ISSUE THREAD

**Issue by cardonator** [zmap/zlint#626]
Hi all, I noticed when attempting to upgrade our internal zlint wrapper to zlint 3.2.0 that zlint is no longer able to compile with the following errors.
```
# github.com/zmap/zlint/v3/util
../../go/pkg/mod/github.com/zmap/zlint/v3@v3.2.0/util/encodings.go:38:27: cannot use atv.Type (type "github.com/zmap/zcrypto/encoding/asn1".ObjectIdentifier) as type "encoding/asn1".ObjectIdentifier in argument to IsNameAttribute
../../go/pkg/mod/github.com/zmap/zlint/v3@v3.2.0/util/oid.go:138:17: cannot use v.Type (type "github.com/zmap/zcrypto/encoding/asn1".ObjectIdentifier) as type "encoding/asn1".ObjectIdentifier in argument to oid.Equal
```
I checked out a later release of zlint and it works so I'm guessing these issues were fixed in a commit back in June. Can we cut a new minor/patch release or something in order to get this working again?

**Comment by robplee** [comment]
Hi @cardonator, my initial thought it that you haven't mentioned which version of golang you're using.  zlint 3.2.0 included a requirement to update your go version from 1.15.x to 1.16.y (can't remember the exact numbers for x and y) so might you be running a version of go too old for zlint 3.2.0?

I've tried and succeeded to import zlint 3.2.0 into a project so I am very much expecting this to be a problem for you only but you never know if it's a wider issue or not :)

**Comment by cardonator** [comment]
Thanks, Rob. I am using Go 1.16.6 and observed the same behavior from a local Go instance (in Linux) versus an Alpine Go docker image on 1.16.6 as well. I haven't tried to test if it has something to do with my wrapper or the wrapper pulling in its own zcrypto dependency, but the errors show up on 3.2.0 and don't appear after the commits on June 20th.

**Comment by dadrian** [comment]
Can you post you `go.sum`? I suspect this is because ZCrypto doesn't have releases, and version 3.2.0 of ZLint is not prepared for commit https://github.com/zmap/zcrypto/commit/12fab66869644acf821e0fd110cdfa40def19fa3 in ZCrypto, which changes the type signature of many functions. When fetching clean with no pins of ZCrypto, Go will fetch the latest version of ZCrypto that is not a major rev as compared to the version in v3.2.0 ZLint's Go mod, which is an unversioned commit hash corresponding to a v0.x. ZCrypto after https://github.com/zmap/zcrypto/commit/12fab66869644acf821e0fd110cdfa40def19fa3 is still a v0.x, Go just fetches ZCrypto HEAD.

To fix, we would need to push out a new release of ZLint that depends on latest ZCrypto, or you need to pin your ZCrypto to prior to https://github.com/zmap/zcrypto/commit/12fab66869644acf821e0fd110cdfa40def19fa3, or you need to pin ZLint to https://github.com/zmap/zlint/commit/dbd9bfd21e5cb6c78c4ba16c87da21492a999c93 or later.

**Comment by cardonator** [comment]
I suspect you're exactly right that that is the issue. The version of zcrypto I'm pinning in my go.mod is `github.com/zmap/zcrypto v0.0.0-20210811211718-6f9bc4aff20f`. I'm not sure how to work with this version format, how do I switch to the commit immediately before the one you mentioned?

If that solves the issue I think it's fine, but it might be worth noting that in the release notes on this project so people avoid the issue in the future. Since zcrypto isn't versioned it makes it hard to rely on the transitive dependency.

**Comment by cardonator** [comment]
After looking into this a bit more, there is a type issue in zlint that's to blame for this. There are some lints in zlint that use the encoding/asn1 type where zcrypto/x509 is expected. In the commit I mentioned above in zlint, the offending lints were all switched to use zcrypto/x509 instead, which fixes the issue. In order to solve this issue or upgrade to a 3.2.0 release or newer we would have to leave the release versioning and pin a specific newer commit of zlint.

**Comment by dadrian** [comment]
I think the easiest thing to do would be to cut a 3.2.1 from HEAD on both projects. I don't know our process for doing so, maybe @christopher-henderson can help us out.

**Comment by christopher-henderson** [comment]
A fix for this is included in the [v3.3.0 Release Candidate](https://github.com/zmap/zlint/releases/tag/v3.3.0-rc1).
